Core/Tornado/Client/private/TornadoBaseClient.py

```diff
@@ -0,0 +1,564 @@
+                        without going through the JEncode serialization.
+
+
+
+      :returns: The received data. If outputFile is set, return always S_OK
+
+    """
+
+    rawContent = kwargs.get('rawContent')
+
+    # Adding some informations to send
+    if self.__extraCredentials:
+      kwargs[self.KW_EXTRA_CREDENTIALS] = encode(self.__extraCredentials)
+    kwargs["clientVO"] = self.vo
+
+    # Getting URL
+    url = self.__findServiceURL()
+    if not url['OK']:
+      return url
+    url = url['Value']
+
+    # Getting CA file (or skip verification)
+    verify = (not self.kwargs[self.KW_SKIP_CA_CHECK])
+    if verify and self.__ca_location:
+      verify = self.__ca_location
+
+    # getting certificate
+    if self.kwargs[self.KW_USE_CERTIFICATES]:
+      cert = Locations.getHostCertificateAndKeyLocation()
+    else:
+      cert = self.__proxy_location
+
+    # Do the request
+    try:
+      rawText = None
+
+      if not stream:
+        call = requests.post(url, data=kwargs,
+                             timeout=self.timeout, verify=verify,
+                             cert=cert)
+        # raising the exception for status here
+        # means essentialy that we are losing here the information of what is returned by the server
+        # as error message, since it is not passed to the exception
+        # However, we can store the text and return it raw as an error,
+        # since there is no guarantee that it is any JEncoded text
+        # Note that we would get an exception only if there is an exception on the server side which
+        # is not handled.
+        # Any standard S_ERROR will be transfered as an S_ERROR with a correct code.
+        rawText = call.text
+        call.raise_for_status()
+        return decode(rawText)[0]
+      else:
+        rawText = None
+        # Stream download
+        # https://requests.readthedocs.io/en/latest/user/advanced/#body-content-workflow
+        with requests.post(url, data=kwargs, timeout=self.timeout, verify=verify,
+                           cert=cert, stream=True) as r:
+          r.raise_for_status()
+          contentFile = outputFile if outputFile else cStringIO.StringIO()
+
+          with open(contentFile, 'wb') as f:
+            for chunk in r.iter_content():
+              if chunk:  # filter out keep-alive new chuncks
+                f.write(chunk)
+
+          if outputFile:
+            return S_OK()
+
+          if rawContent:
+            return S_OK(contentFile.getvalue())
+          return decode(contentFile.getvalue())[0]
+
+    except Exception as e:
+      # CHRIS TODO review this part: catch specific exceptions
+      # self.__bannedUrls is emptied in findServiceURLs
+      if url not in self.__bannedUrls:
+        self.__bannedUrls += [url]
+      if retry < self.__nbOfUrls - 1:
+        self._request(retry=retry + 1, stream=stream, outputFile=outputFile, **kwargs)
+
+      errStr = "%s: %s" % (str(e), rawText)
+      return S_ERROR(errStr)
```

Comment:
It may be worth here returning `S_ERROR` with the response status code or describe the http codes in the `DErrno` and return them according to the received status code.

Reply:
Hum, I think I understand what you mean. I'll think of it, but I do not think it is a good idea to re-map all the HTTP status code to a `DErrno` error code. However you are correct that we should distinguish the cases where we get really an HTTP error, or if we get another exception (network for example). I'll try to come up with something